ocaml/xcp-rrdd/bin/rrdd/rrdd_server.ml

```diff
@@ -456,6 +457,26 @@ let query_host_ds (ds_name : string) : float =
           query ds_name rrdi
   )
 
+(** Dump all latest data of host dss to file in json format so that any client
+    can read even if it's non-privileged user, such as NRPE. *)
+let dump_host_dss_to_file (file : string) : unit =
+  let json =
+    with_lock mutex (fun () ->
+        match !host_rrd with
+        | None ->
+            `Assoc []
+        | Some rrdi ->
+            `Assoc
+              (Rrd.ds_names rrdi.rrd
+              |> List.map (fun ds_name -> (ds_name, `Float (query ds_name rrdi)))
+              )
+    )
+  in
+  let oc = open_out_gen [Open_wronly; Open_creat; Open_trunc] 0o644 file in
+  finally
+    (fun () -> Yojson.Basic.pretty_to_channel oc json)
```

Comment:
We probably want a compact representation after the testing is done:
```suggestion
    (fun () -> Yojson.Basic.to_channel oc json)
```

Reply:
I saw `query ds_name rrdi` already ensure return `float` type value, so I still need to handle NaN and infinities cases for Yojson?

Reply:
Unfortunately, yes. Json doesn't allow NaN not infinities to be represented in floats, but NaN in this case must be considered a common value. I don't think infinities appear hear, but they are a possibility.

https://datatracker.ietf.org/doc/html/rfc7159.html#section-6
> Numeric values that cannot be represented in the grammar below (such
>   as Infinity and NaN) are not permitted.

>      number = [ minus ] int [ frac ] [ exp ]
>      decimal-point = %x2E       ; .
>      digit1-9 = %x31-39         ; 1-9
>      e = %x65 / %x45            ; e E
>      exp = e [ minus / plus ] 1*DIGIT
>      frac = decimal-point 1*DIGIT
>      int = zero / ( digit1-9 *DIGIT )
>      minus = %x2D               ; -
>      plus = %x2B                ; +
>      zero = %x30                ; 0

>   This specification allows implementations to set limits on the range
>   and precision of numbers accepted.  Since software that implements
>   IEEE 754-2008 binary64 (double precision) numbers [[IEEE754](https://datatracker.ietf.org/doc/html/rfc7159.html#ref-IEEE754)] is
>   generally available and widely used, good interoperability can be
>   achieved by implementations that expect no more precision or range
>   than these provide, in the sense that implementations will
>   approximate JSON numbers within the expected precision.  A JSON
>   number such as 1E400 or 3.141592653589793238462643383279 may indicate
>   potential interoperability problems, since it suggests that the
>   software that created it expects receiving software to have greater
>   capabilities for numeric magnitude and precision than is widely
>   available.

>   Note that when such software is used, numbers that are integers and
>   are in the range [-(2**53)+1, (2**53)-1] are interoperable in the
>   sense that implementations will agree exactly on their numeric
>   values.

I believe some sort of agreement between writer and reader must be done in order to parse these 3 values (NaN, Infinity, -Infinity)

Reply:
When using std=true whenever NaN or the infinities are present the library throws an exception. They need some treatment in the representation before they can be sent. This is what I meant by
> some sort of agreement between writer and reader must be done in order to parse these 3 values 
The representation must be agreed, documented in the design, then implemented in both sides

Reply:
Fixed now, could you continue review?
I also updated design https://info.citrite.net/display/xenserver/Design+of+NRPE+integration and commit description.

Reply:
Is it worth to add some unit tests for these agreements?

Reply:
This is simple and not critical function, just let me know if UT is must have.